state-sync/storage-service/server/src/lib.rs

```diff
@@ -166,25 +373,144 @@ impl<T: StorageReaderInterface> StorageServiceServer<T> {
+) -> Result<Vec<(AccountAddress, LedgerInfoWithSignatures)>, Error> {
+    // Fetch the latest storage summary and highest synced version
+    let latest_storage_summary = cached_storage_server_summary.read().clone();
+    let highest_synced_ledger_info = match latest_storage_summary.data_summary.synced_ledger_info {
+        Some(ledger_info) => ledger_info,
+        None => return Ok(vec![]),
+    };
+    let highest_synced_version = highest_synced_ledger_info.ledger_info().version();
+    let highest_synced_epoch = highest_synced_ledger_info.ledger_info().epoch();
+
+    // Identify the peers with ready subscriptions
+    let mut ready_subscriptions = vec![];
+    for (peer, data_subscription) in data_subscriptions.lock().iter() {
+        if data_subscription.highest_known_version() < highest_synced_version {
+            let highest_known_epoch = data_subscription.highest_known_epoch();
+            let target_ledger_info = if highest_known_epoch < highest_synced_epoch {
+                // The peer needs to sync to their epoch ending ledger info
+                get_epoch_ending_ledger_info(highest_known_epoch, storage.clone())?
+            } else {
+                highest_synced_ledger_info.clone()
+            };
+            ready_subscriptions.push((*peer, target_ledger_info));
+        }
+    }
+    Ok(ready_subscriptions)
+}
+
+/// Gets the epoch ending ledger info at the given epoch
+fn get_epoch_ending_ledger_info<T: StorageReaderInterface>(
+    epoch: u64,
+    storage: T,
+) -> Result<LedgerInfoWithSignatures, Error> {
+    match storage.get_epoch_ending_ledger_infos(epoch, epoch) {
+        Ok(epoch_change_proof) => {
+            if let Some(ledger_info) = epoch_change_proof.ledger_info_with_sigs.get(0) {
+                Ok(ledger_info.clone())
+            } else {
+                Err(Error::StorageErrorEncountered(
+                    "Empty change proof found!".into(),
+                ))
+            }
+        }
+        Err(error) => Err(Error::StorageErrorEncountered(format!(
+            "Failed to get epoch ending ledger info! Error: {:?}",
+            error
+        ))),
+    }
+}
+
+/// Notifies a subscriber of new data according to the target ledger info
+fn notify_peer_of_new_data<T: StorageReaderInterface>(
+    cached_storage_server_summary: Arc<RwLock<StorageServerSummary>>,
+    config: StorageServiceConfig,
+    data_subscriptions: Arc<Mutex<HashMap<AccountAddress, DataSubscriptionRequest>>>,
+    lru_storage_cache: Arc<Mutex<LruCache<StorageServiceRequest, StorageServiceResponse>>>,
+    storage: T,
+    time_service: TimeService,
+    subscription: DataSubscriptionRequest,
+    target_ledger_info: LedgerInfoWithSignatures,
+) -> Result<(), Error> {
+    match subscription.get_storage_request_for_missing_data(config, &target_ledger_info) {
+        Ok(storage_request) => {
+            // Handle the storage service request to fetch the missing data
+            let handler = Handler::new(
```

Comment:
it's weird to create handler each time per notification

Reply:
Yeah, this is primarily just for code reuse (e.g., so the data fetch goes through the cache and the appropriate error handling/logging/metrics, etc). Also, by creating the handler we can technically spawn this on a separate thread (e.g., if we see we're getting a lot of subscriptions). I'm inclined to leave it as-is. Otherwise, the refactor will be huge.